config/config_test.go

```diff
@@ -71,54 +72,229 @@ func TestCreateConfig(t *testing.T) {
 			},
-			filePath: cnfPath,
 		}
 
-		actual, err := New(cnfPath)
+		actual, err := configList.Get("default")
+		if err != nil {
+			t.Fatal(err)
+		}
 
 		assert.Nil(t, err)
-		assert.Equal(t, cnfPath, actual.filePath)
 		assert.Equal(t, testConfig, actual)
 	})
 
 	t.Run("failure by invalid current", func(t *testing.T) {
-		cnfPath := filepath.Join(testDir, "failureCurrentConfig")
-		_, err := New(cnfPath)
+		configList := ConfigList{
+			Configs: map[string]*Config{
+				"default": {
+					APIURL:   "api-url",
+					StoreURL: "store-api-url",
+					Token:    "dummy_token",
+					Launcher: Launcher{
+						Version: "latest",
+						Image:   "screwdrivercd/launcher",
+					},
+				},
+			},
+			Current: "doesnotexist",
+		}
+
+		_, err := configList.Get(configList.Current)
 
 		assert.Equal(t, "config `doesnotexist` does not exist", err.Error())
 	})
+}
 
-	t.Run("failure by invalid yaml", func(t *testing.T) {
-		cnfPath := filepath.Join(testDir, "failureConfig")
-		_, err := New(cnfPath)
+func TestConfigListAdd(t *testing.T) {
+	t.Run("success", func(t *testing.T) {
+		configList := ConfigList{
+			Current: "default",
+			Configs: map[string]*Config{
+				"default": {
+					APIURL:   "api-url",
+					StoreURL: "store-api-url",
+					Token:    "dummy_token",
+					Launcher: Launcher{
+						Version: "latest",
+						Image:   "screwdrivercd/launcher",
+					},
+				},
+			},
+		}
 
-		assert.Equal(t, 0, strings.Index(err.Error(), "failed to parse config file: "), fmt.Sprintf("expected error is `failed to parse config file: ...`, actual: `%v`", err))
+		err := configList.Add("test")
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		expected := ConfigList{
+			Configs: map[string]*Config{
+				"default": {
+					APIURL:   "api-url",
+					StoreURL: "store-api-url",
+					Token:    "dummy_token",
+					Launcher: Launcher{
+						Version: "latest",
+						Image:   "screwdrivercd/launcher",
+					},
+				},
+				"test": {
+					APIURL:   "",
+					StoreURL: "",
+					Token:    "",
+					Launcher: Launcher{
+						Version: "stable",
+						Image:   "screwdrivercd/launcher",
+					},
+				},
+			},
+			Current: "default",
+		}
+
+		assert.Equal(t, expected, configList)
+	})
+
+	t.Run("failure by the name that exists", func(t *testing.T) {
+		configList := ConfigList{
+			Current: "default",
+			Configs: map[string]*Config{
+				"default": {
+					APIURL:   "api-url",
+					StoreURL: "store-api-url",
+					Token:    "dummy_token",
+					Launcher: Launcher{
+						Version: "latest",
+						Image:   "screwdrivercd/launcher",
+					},
+				},
+			},
+		}
+
+		err := configList.Add("default")
+		assert.Equal(t, "config `default` already exists", err.Error())
 	})
 }
 
-func TestSetConfig(t *testing.T) {
-	rand.Seed(time.Now().UnixNano())
-	cnfPath := filepath.Join(testDir, ".sdlocal", fmt.Sprintf("%vconfig", rand.Int()))
-	defer os.RemoveAll(filepath.Join(testDir, ".sdlocal"))
+func TestConfigListSave(t *testing.T) {
+	t.Run("success", func(t *testing.T) {
```

Comment:
Is there a good way to test a failing case?

Reply:
It looks tough to do failure case tests of Save function.
I think we do not need the failure case.

Reply:
Thx.
I will not make it.